test/ha/resilience_test.go

```diff
@@ -0,0 +1,169 @@
+			autoscaling.MaxScaleAnnotationKey:  strconv.Itoa(maximumNumberOfReplicas),
+			autoscaling.TargetBurstCapacityKey: "0", // The Activator is only added to the request path during scale from zero scenarios.
+		}),
+	)
+	test.EnsureTearDown(t, clients, &names)
+
+	testUptimeDuringUserPodDeletion(t, ctx, clients, names, resources)
+}
+
+func TestActivatorInRequestPathAlways(t *testing.T) {
+	clients := e2e.Setup(t)
+	ctx := context.Background()
+
+	// Create first service that we will continually probe during disruption scenario.
+	names, resources := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.MinScaleAnnotationKey:  strconv.Itoa(minimumNumberOfReplicas), // Make sure we don't scale to zero during the test.
+			autoscaling.MaxScaleAnnotationKey:  strconv.Itoa(maximumNumberOfReplicas),
+			autoscaling.TargetBurstCapacityKey: "-1", // Make sure all requests go through the activator.
+		}),
+	)
+	test.EnsureTearDown(t, clients, &names)
+
+	testUptimeDuringUserPodDeletion(t, ctx, clients, names, resources)
+}
+
+func TestActivatorInRequestPathPossibly(t *testing.T) {
+	clients := e2e.Setup(t)
+	ctx := context.Background()
+
+	// Create first service that we will continually probe during disruption scenario.
+	names, resources := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.MinScaleAnnotationKey:  strconv.Itoa(minimumNumberOfReplicas), // Make sure we don't scale to zero during the test.
+			autoscaling.MaxScaleAnnotationKey:  strconv.Itoa(maximumNumberOfReplicas),
+			autoscaling.TargetBurstCapacityKey: "1", // The Activator may be in the path, depending on the revision scale and load.
+		}),
+	)
+	test.EnsureTearDown(t, clients, &names)
+
+	testUptimeDuringUserPodDeletion(t, ctx, clients, names, resources)
+}
+
+func testUptimeDuringUserPodDeletion(t *testing.T, ctx context.Context, clients *test.Clients, names test.ResourceNames, resources *v1test.ResourceObjects) {
+	t.Log("Starting prober")
+	prober := test.NewProberManager(t.Logf, clients, minProbes, test.AddRootCAtoTransport(context.Background(), t.Logf, clients, test.ServingFlags.HTTPS))
+
+	prober.Spawn(resources.Service.Status.URL.URL())
+	defer assertSLO(t, prober, 1)
+
+	// Get user pods
+
+	selector := labels.SelectorFromSet(labels.Set{
+		serving.ServiceLabelKey: names.Service,
+	})
+	pods, err := clients.KubeClient.CoreV1().Pods(test.ServingFlags.TestNamespace).List(ctx, metav1.ListOptions{LabelSelector: selector.String()})
+	if err != nil {
+		t.Fatalf("Unable to get pods: %v", err)
+	}
+
+	if !(len(pods.Items) == minimumNumberOfReplicas) {
+		t.Fatalf("Expected to have %d user pod(s) running, but found %d.", minimumNumberOfReplicas, len(pods.Items))
+	}
+
+	t.Logf("Watching user pods")
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go watchPodEvents(t, ctx, clients, &wg, selector, pods.Items[0].Name)
+	go watchPodEvents(t, ctx, clients, &wg, selector, pods.Items[1].Name)
+	wg.Wait()
```

Comment:
Can we do this numerous times in the test - say like 10 times to just flush out that traffic doesn't get dropped

Reply:
Okay, understood. How about the deletion coordination process? Currently, we attempt to delete both pods simultaneously and then wait for confirmation that both have been deleted before proceeding with the assertions. Is this approach reasonable, or should we delete one pod, ensure it's successfully deleted, and then proceed with deleting the next? (with 10 iterations, the test time will increase for sure)

Reply:
let's do one pod deletion at time - if we assume the default distribution is a pod on different nodes then that would mimic the draining